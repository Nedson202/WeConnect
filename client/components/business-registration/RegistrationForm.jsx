import React from 'react';
import PropTypes from 'prop-types';
/**
 * @class RegistrationForm
 *
 * @extends {Component}
 */
const RegistrationForm = ({
  params,
  categories,
  locations,
  state,
  onChange,
  onSubmit,
  isLoading,
  changeValue, 
  defaultTotal, 
  color
}) => {
  // const { changeValue, defaultTotal, color } = state;

  // const categoryOption = categories.map(({ id, category }) =>
  //   <option key={id} value={category}>{category}</option>);

  // const locationOption = locations.map(({ id, location }) =>
  //   <option key={id} value={location}>{location}</option>);

  /**
   * @description Renders the component to the dom
   *
   * @returns {object} JSX object
   *
   * @memberof RegistrationForm
   */
  return (
    <div className="container">
      <form onSubmit={onSubmit}>
        <div className="business-form-style">
          <h3 className="text-center form-header">
            { !params.id ? 'Register business' : 'Update business' }
          </h3>
          <div className="row">
            <div className="form-group col-lg-6">
              <label id="control-label">Business Name</label>
              <input
                value={state.name}
                onChange={onChange}
                type="text"
                name="name"
                className="form-control"
                id="control-label"
                placeholder="Business name"
              />
            </div>
            <div className="form-group col-lg-6">
              <label id="control-label">Email address</label>
              <input
                value={state.email}
                onChange={onChange}
                type="email"
                name="email"
                id="icon-prefix email"
                className="form-control"
                aria-describedby="emailHelp"
                placeholder="Enter email"
              />
            </div>
          </div>
          <div className="row">
            <div className="form-group col-lg-6">
              <label id="control-label">Address</label>
              <input
                value={state.address}
                onChange={onChange}
                type="text"
                name="address"
                className="form-control"
                id="control-label"
                placeholder="Address"
              />
            </div>
            <div className="form-group col-lg-6">
              <label id="control-label">Location</label>
              <select
                className="form-control custom-select"
                type="select"
                name="location"
                onChange={onChange}
                value={state.location}
              >
                <option value="" disabled>choose location</option>
                {locations()}
              </select>
            </div>
          </div>
          <div className="row">
            <div className="form-group col-lg-6">
              <label id="control-label">Category</label>
              <select
                className="form-control custom-select"
                type="select"
                name="category"
                onChange={onChange}
                value={state.category}
              >
                <option value="" disabled>choose category</option>
                {categories ()}
              </select>
            </div>
            <div className="form-group col-lg-6">
              <label id="control-label">Description</label>
              <textarea
                value={state.description}
                onChange={onChange}
                type="text"
                name="description"
                className="form-control"
                id="control-label"
                placeholder="Description"
                rows="3"
              />
              <p><span style={{ color }}>{changeValue}</span><span style={{ color: 'white' }}>/{defaultTotal}</span></p>
            </div>
          </div>
          <button type="submit" className="btn btn-outline-success" id="submit-button">
            {isLoading ? <span className="processing-info">processing <i className="fa fa-spinner fa-spin" /></span> :
            <span><i className="fa fa-paper-plane" />&nbsp;
              { !params.id ? 'Submit' : 'Update' }
            </span>}
          </button>
        </div>
      </form>
    </div>
  );
};

RegistrationForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  state: PropTypes.object.isRequired,
  params: PropTypes.object.isRequired,
  categories: PropTypes.func.isRequired,
  locations: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  changeValue: PropTypes.number.isRequired,
  defaultTotal: PropTypes.number.isRequired,
  color: PropTypes.string.isRequired,
};

export default RegistrationForm;
