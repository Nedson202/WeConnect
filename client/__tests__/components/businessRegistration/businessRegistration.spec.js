import React from 'react';
import { configure, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import toJSON from 'enzyme-to-json';
import ConnectedBusinessRegistration,
{ BusinessRegistration }
  from '../../../components/business-registration/BusinessRegistration';
import RegistrationForm from '../../../components/business-registration/RegistrationForm';
import state from '../../__mockData__/business';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const { business, categories, locations } = state;

configure({ adapter: new Adapter() });
let props;
props = {
  businessRegistrationRequest: jest.fn(() => Promise.resolve()),
  businessUpdateRequest: jest.fn(() => Promise.resolve()),
  fetchBusinessById: jest.fn(() => Promise.resolve()),
  fetchCategories: jest.fn(() => Promise.resolve()),
  categories: jest.fn(() => Promise.resolve()),
  fetchLocations: jest.fn(() => Promise.resolve()),
  locations: jest.fn(() => Promise.resolve()),
  isLoading: false,
  match: {
    params: {
      id: 1
    }
  },
  params: {
    id: 1
  },
  history,
  categories,
  locations,
  onChange: jest.fn(() => Promise.resolve()),
  onSubmit: jest.fn(() => Promise.resolve()),
  state: {
    businesses: {
      categories: () => [],
      locations: () => []
    },
    business
  }
};
const setup = () => shallow(<BusinessRegistration {...props} />);

const event = (name, value) => ({
  preventDefault: jest.fn(() => Promise.resolve()),
  target: {
    name,
    value
  }
});

describe('state to props', () => {
  props = {
    match: {
      params: {}
    },
  };

  it('should render connected registration page', () => {
    const businesses = {
      business,
      params,
      categories,
      locations,
    };

    const store = mockStore({ businesses });
    const wrapper = shallow(<ConnectedBusinessRegistration store={store} {...props} />);
    expect(wrapper.length).toBe(1);
  });
});

describe('componentWillReceiveProps()', () => {
  const nextProps = {
    business
  };
  it('should return next props', () => {
    const wrapper = shallow(<BusinessRegistration {...props} />);
    const action = wrapper.instance();
    const componentWillReceiveProps = jest.spyOn(wrapper.instance(), 'componentWillReceiveProps');
    action.componentWillReceiveProps(nextProps);
    expect(componentWillReceiveProps).toBeCalled();
  });
});

describe('Business registration component', () => {
  beforeEach(() => {
    global.props = {
      match: {
        params: () => {}
      },
      state: {
        businesses: {
          business: () => {},
          categories: () => {}
        }
      }
    };
  });

  it('should render registration page', () => {
    const wrapper = setup();
    expect(wrapper.find('div').length).toBe(2);
    expect(wrapper.find('RegistrationForm').length).toBe(1);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('should render registration form', () => {
    props = {
      state: {
        name: () => {}
      },
      locations: jest.fn(() => Promise.resolve()),
      categories: jest.fn(() => Promise.resolve()),
      onChange: jest.fn(() => Promise.resolve()),
      onSubmit: jest.fn(() => Promise.resolve()),
      params: {
        id: 1
      },
      isLoading: false
    };

    const state = {
      business: null,
      params: { id: 1 },
      categories,
      locations,
      isLoading: false,
    };

    const wrapper = shallow(<RegistrationForm {...props} />);
    expect(wrapper.find('.form-header').text()).toBe('Update business');
    wrapper.setProps({ params: {} });
    expect(wrapper.find('.form-header').text()).toBe('Register business');
    expect(wrapper.find('div').length).toBe(11);
    expect(wrapper.find('form').length).toBe(1);
    expect(wrapper.find('button').length).toBe(1);
    expect(wrapper.find('i').length).toBe(1);
    expect(wrapper.find('textarea').length).toBe(1);
    expect(wrapper.find('span').length).toBe(3);
    expect(wrapper.find('button').simulate('click'));
    wrapper.setProps({ isLoading: true, params: { id: 1 } });
    expect(wrapper.find('.form-header').text()).toBe('Update business');
    expect(wrapper.find('span').length).toBe(3);
    expect(toJSON(wrapper)).toMatchSnapshot();


    props = {
      match: {
        params: {
          // id: 1
        }
      },
    };

    // expect(mapStateToProps(state, props)).toEqual(state);
  });

  it('should render connected registration page', () => {
    const businesses = {
      business,
      params,
      categories,
      locations,
    };

    const store = mockStore({ businesses });
    const wrapper = shallow(<ConnectedBusinessRegistration store={store} {...props} />);
    expect(wrapper.length).toBe(1);
  });
});

describe('Registration onchange method', () => {
  beforeEach(() => {
    global.document = {
      title: () => {}
    };
  });


  it('should simulate input change', () => {
    const wrapper = setup();
    const RegistrationForm = wrapper.find('RegistrationForm');

    RegistrationForm.simulate('change', event('name', business.name));
    expect(wrapper.instance().state.name).toBe(business.name);

    RegistrationForm.simulate('change', event('email', business.email));
    expect(wrapper.instance().state.email).toBe(business.email);

    RegistrationForm.simulate('change', event('address', business.address));
    expect(wrapper.instance().state.address).toBe(business.address);

    RegistrationForm.simulate('change', event('location', business.location));
    expect(wrapper.instance().state.location).toBe(business.location);

    RegistrationForm.simulate('change', event('category', business.category));
    expect(wrapper.instance().state.category).toBe(business.category);

    RegistrationForm.simulate('change', event('description', business.description));
    expect(wrapper.instance().state.description).toBe(business.description);
  });

  it('should simulate password change', () => {
    const wrapper = setup();
    const RegistrationForm = wrapper.find('RegistrationForm');
  });
});

describe('Submit function', () => {
  it('should submit business information', () => {
    const wrapper = setup();
    const RegistrationForm = wrapper.find('RegistrationForm');
    wrapper.setState(business);

    RegistrationForm.simulate('submit', event());
  });
});

describe('Categories and location', () => {
  it('should call the get locations function', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const location = jest.spyOn(wrapper.instance(), 'mapLocations');
    action.mapLocations();
    expect(location).toBeCalled();
  });

  it('should call the get categories function', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const categories = jest.spyOn(wrapper.instance(), 'mapCategories');
    action.mapCategories();
    expect(categories).toBeCalled();
  });
});

