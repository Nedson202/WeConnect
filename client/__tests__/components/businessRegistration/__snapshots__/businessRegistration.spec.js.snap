// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Business registration component should render registration form 1`] = `
<div
  className="container"
>
  <form
    onSubmit={[MockFunction]}
  >
    <div
      className="business-form-style"
    >
      <h3
        className="text-center form-header"
      >
        Update business
      </h3>
      <div
        className="row"
      >
        <div
          className="form-group col-lg-6"
        >
          <label
            id="control-label"
          >
            Business Name
          </label>
          <input
            className="form-control"
            id="control-label"
            name="name"
            onChange={[MockFunction]}
            placeholder="Business name"
            type="text"
            value={[Function]}
          />
        </div>
        <div
          className="form-group col-lg-6"
        >
          <label
            id="control-label"
          >
            Email address
          </label>
          <input
            aria-describedby="emailHelp"
            className="form-control"
            id="icon-prefix email"
            name="email"
            onChange={[MockFunction]}
            placeholder="Enter email"
            type="email"
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="form-group col-lg-6"
        >
          <label
            id="control-label"
          >
            Address
          </label>
          <input
            className="form-control"
            id="control-label"
            name="address"
            onChange={[MockFunction]}
            placeholder="Address"
            type="text"
          />
        </div>
        <div
          className="form-group col-lg-6"
        >
          <label
            id="control-label"
          >
            Location
          </label>
          <select
            className="form-control custom-select"
            name="location"
            onChange={[MockFunction]}
            type="select"
          >
            <option
              disabled={true}
              value=""
            >
              choose location
            </option>
          </select>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="form-group col-lg-6"
        >
          <label
            id="control-label"
          >
            Category
          </label>
          <select
            className="form-control custom-select"
            name="category"
            onChange={[MockFunction]}
            type="select"
          >
            <option
              disabled={true}
              value=""
            >
              choose category
            </option>
          </select>
        </div>
        <div
          className="form-group col-lg-6"
        >
          <label
            id="control-label"
          >
            Description
          </label>
          <textarea
            className="form-control"
            id="control-label"
            name="description"
            onChange={[MockFunction]}
            placeholder="Description"
            rows="3"
            type="text"
          />
          <p>
            <span
              style={
                Object {
                  "color": undefined,
                }
              }
            />
            <span
              style={
                Object {
                  "color": "white",
                }
              }
            >
              /
            </span>
          </p>
        </div>
      </div>
      <button
        className="btn btn-outline-success"
        id="submit-button"
        type="submit"
      >
        <span
          className="processing-info"
        >
          processing 
          <i
            className="fa fa-spinner fa-spin"
          />
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`Business registration component should render registration page 1`] = `
<div>
  <div
    id="background-image"
  />
  <RegistrationForm
    categories={[Function]}
    changeValue={0}
    color="red"
    defaultTotal={250}
    isLoading={false}
    locations={[Function]}
    onChange={[Function]}
    onSubmit={[Function]}
    params={
      Object {
        "id": 1,
      }
    }
    state={
      Object {
        "address": undefined,
        "category": undefined,
        "changeValue": 0,
        "color": "red",
        "defaultTotal": 250,
        "description": undefined,
        "email": undefined,
        "errors": Object {},
        "location": undefined,
        "name": undefined,
      }
    }
  />
</div>
`;
